File: C++/InversePairs.cpp
#include<iostream>
#include<vector>

class Solution{
    public:
    int InversePairs(std::vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        std::vector<int> temp(nums.size());
        return mergeSortCount(nums, temp, 0, nums.size() - 1);
    }

    int mergeSortCount(std::vector<int>& nums, std::vector<int>& temp, int start, int end) {
        if (start >= end) {
            return 0;
        }
        int mid = start + (end - start) / 2;
        int countLeft = mergeSortCount(nums, temp, start, mid);
        int countRight = mergeSortCount(nums, temp, mid + 1, end);
        int mergeCountResult = mergeCount(nums, temp, start, mid, end);
        return countLeft + countRight + mergeCountResult;
    }

    int mergeCount(std::vector<int>& nums, std::vector<int>& temp, int start, int mid, int end) {
        int count = 0;
        int i = mid, j = end, k = end; 
        while (i >= start && j > mid) {
            if (nums[i] > nums[j]) {
                count += (j - mid);
                temp[k--] = nums[i--];
            } else {
                temp[k--] = nums[j--];
            }
        }
        while (i >= start) {
            temp[k--] = nums[i--];
        }
        while (j > mid) {
            temp[k--] = nums[j--];
        }
        for (int m = start; m <= end; ++m) {
            nums[m] = temp[m];
        }
        return count;
    }
};

int main(){
    std::vector<int>nums={1,2,3,4,5,6,7,0};
    Solution solution;
    int Inversecount=solution.InversePairs(nums);
    std::cout<<Inversecount<<std::endl;
    return 0;
}

File: python/InversePairs.py
class Solution:
    mod = 1000000007
    def InversePairs(self, nums: list[int]) -> int:
        n=len(nums)
        res = [0 for i in range(n)]
        return self.MergeSort(0,n-1,nums,res)
    def MergeSort(self,left:int,right:int,nums:list[int],temp:list[int])-> int:
        if left >=right:
            return 0
        mid= (left+right)//2
        res = self.MergeSort(left,mid,nums,temp) + self.MergeSort(mid+1,right,nums,temp)
        res %=self.mod
        i=left
        j=mid+1
        for k in range(left,right+1):
            temp[k]=nums[k]
        for k in range (left,right+1):
            if i == mid+1:
                nums[k]=temp[j]
                j +=1
            elif j==right +1 or temp[i]<=temp[j]:
                nums[k]=temp[i]
                i +=1
            else:
                nums[k]=temp[j]
                j +=1
                res +=mid-i+1
        return res % self.mod

nums=[1,2,3,4,5,6,7,0]
solution=Solution()
print(solution.InversePairs(nums))